shark-android/src/test/java/shark/AndroidReferenceMatcher_XIAMI__RESOURCES_IMPL_Test.kt

```diff
@@ -0,0 +1,173 @@
+package shark
+
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.Test
+import java.io.File
+import shark.AndroidReferenceMatchers.Companion.HMD_GLOBAL
+import shark.AndroidReferenceMatchers.Companion.INFINIX
+import shark.AndroidReferenceMatchers.Companion.LENOVO
+import shark.AndroidReferenceMatchers.Companion.NVIDIA
+import shark.AndroidReferenceMatchers.Companion.XIAOMI
+import shark.HprofHeapGraph.Companion.openHeapGraph
+import shark.ReferencePattern.StaticFieldPattern
+import shark.ValueHolder.IntHolder
+
+/**
+ * Tests for [AndroidReferenceMatchers.XIAMI__RESOURCES_IMPL] known leaks, to ensure we
+ * don't report the ones we know about as application leaks, and do report the ones we don't.
+ */
+class AndroidReferenceMatcher_XIAMI__RESOURCES_IMPL_Test {
+
+  /**
+   * It's known that some device manufacturers on SDK 30 and above leak
+   * a statically held [android.content.Context] from a class [android.content.res.ResourcesImpl]
+   * in a variable named [mContext].
+   */
+  private val expectedKnownClassLeakingContext = "android.content.res.ResourcesImpl"
+  private val expectedKnownReferenceName = "mAppContext"
+
+  /**
+   * Arbitrarily stops at 60, but will we really be making Android apps by then?
+   * #AI-will-take-over
+   */
+  @Test fun `Expect anything under SDK 30 unknown`() {
+    for(i in 25..60) {
+      givenLeakingContextFor(HMD_GLOBAL, i).let { heapAnalysis ->
+        when {
+          i < 30 -> heapAnalysis.expectApplicationLeak()
+          else -> heapAnalysis.expectKnownLibraryLeak()
+        }
+      }
+    }
+  }
+
+  /**
+   * NVIDIA is not in the list of known manufacturers, so we would expect an application
+   * leak to be reported.
+   */
+  @Test fun `Expect NVIDIA unknown`() {
+    givenLeakingContextFor(NVIDIA, 30).expectApplicationLeak()
+  }
+
+  @Test fun `Expect XIAOMI known`() {
+    givenLeakingContextFor(XIAOMI, 30).expectKnownLibraryLeak()
+  }
+
+  @Test fun `Expect LENOVO known`() {
+    givenLeakingContextFor(LENOVO, 30).expectKnownLibraryLeak()
+  }
+
+  @Test fun `Expect INFINIX known`() {
+    givenLeakingContextFor(INFINIX, 30).expectKnownLibraryLeak()
+  }
+
+  @Test fun `Expect HMD_GLOBAL known`() {
+    givenLeakingContextFor(HMD_GLOBAL, 30).expectKnownLibraryLeak()
+  }
```

Comment:
These are cool but might want to be a single parameterized test?

Reply:
I always go back and forth on these types of things.

You mean something like


```
@Test fun `Known device manufaturers are set as known lib leaks`() {
    val sdkInt = 30
    val listOf(HMD_GLOBAL, ... ). forEach { manufacturer ->
        givenLeakingContextFor(manufacturer, sdkInt).expectKnownLibraryLeak()
    }
  }
```

Reply:
That's one option, another one is to use junit support for parameterized tests. The benefit is that it does run all cases and reports each as a distinct test.

Reply:
I like my original one better.  I'll push as a new commit, the parameterized version so we can compare. 

If you like it, I can squash before merging.

What I don't like about it, is that the tests run don't tell me what was tested, just an index of the test.

For example : @Parameterized Test Version
<img width="800" alt="image" src="https://github.com/square/leakcanary/assets/219132/56766eab-2960-46dd-9712-18b1c8617b19">

Original, manual Parameterized test version
<img width="800" alt="image" src="https://github.com/square/leakcanary/assets/219132/2a3cc51d-5feb-45e8-8137-5c015630c998">

Reply:
Original approach is fine too

Reply:
Ultimately, I finally decided on the parameterized approach after all because I the scenarios are easier to in a list like that. 🙃